### PR TITLE
[Manual Backport 2.x] [Workspace] Add user settings

### PR DESCRIPTION
**Pull request by Hailong-am**
backport 569b70d885cec37ac83d133ef6d8ce0f6c03a02f from #7953

### CODE REVIEW

**Comment by Hailong-am** [comment]
cypress failure is not related to this change.

**Review by SuZhou-Joe — APPROVED** [review]

File: src/core/server/ui_settings/utils.ts [hunk 1]
```diff
@@ -0,0 +1,15 @@
+/*
+ * Copyright OpenSearch Contributors
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+import { UiSettingScope } from './types';
+
+export const CURRENT_USER_PLACEHOLDER = '<current_user>';
+
+export const generateDocId = (id: string, scope?: UiSettingScope) => {
```
Comment (AMoo-Miki) [inline src/core/server/ui_settings/utils.ts:10]: This function name is not inline with what it is actually doing.

File: src/plugins/advanced_settings/server/saved_objects/user_ui_settings_client_wrapper.ts [hunk 2]
```diff
@@ -0,0 +1,142 @@
...
+
+    const updateUiSettings = async <T = unknown>(
+      type: string,
+      id: string,
+      attributes: Partial<T>,
+      options: SavedObjectsUpdateOptions = {}
+    ): Promise<SavedObjectsUpdateResponse<T>> => {
+      if (type === 'config') {
+        const docId = this.normalizeDocId(id, wrapperOptions.request, this.core);
+        this.logger.debug(`Getting config with original: ${id} normalizeDocId: ${docId}`);
```
Comment (AMoo-Miki) [inline src/plugins/advanced_settings/server/saved_objects/user_ui_settings_client_wrapper.ts:79]: Plz remove the dev logging.

File: src/plugins/advanced_settings/server/saved_objects/user_ui_settings_client_wrapper.ts [hunk 3]
```diff
@@ -0,0 +1,142 @@
...
+
+  public wrapperFactory: SavedObjectsClientWrapperFactory = (wrapperOptions) => {
+    const getUiSettings = async <T = unknown>(
+      type: string,
+      id: string,
+      options: SavedObjectsBaseOptions = {}
+    ): Promise<SavedObject<T>> => {
+      if (type === 'config') {
+        const docId = this.normalizeDocId(id, wrapperOptions.request, this.core);
+        this.logger.debug(`Getting config with original: ${id} normalizeDocId: ${docId}`);
```
Comment (AMoo-Miki) [inline src/plugins/advanced_settings/server/saved_objects/user_ui_settings_client_wrapper.ts:63]: Plz remove the dev logging.

File: src/plugins/workspace/public/components/workspace_list/index.tsx [hunk 4]
```diff
@@ -198,6 +234,22 @@ export const WorkspaceList = ({ registeredUseCases$ }: WorkspaceListProps) => {
     [application, http]
   );
 
+  const handleSetDefaultWorkspace = useCallback(
+    async (item: WorkspaceAttribute) => {
+      const set = await uiSettings?.set(DEFAULT_WORKSPACE, item.id);
+      if (set) {
+        setDefaultWorkspaceId(item.id);
+        notifications?.toasts.addSuccess(`Default workspace been set to ${item.name}`);
```
Comment (AMoo-Miki) [inline src/plugins/workspace/public/components/workspace_list/index.tsx:242]: Needs to use i18n

File: src/plugins/workspace/public/components/workspace_list/index.tsx [hunk 5]
```diff
@@ -187,6 +222,7 @@ export const WorkspaceList = ({ registeredUseCases$ }: WorkspaceListProps) => {
 
   const handleCopyId = (id: string) => {
     copyToClipboard(id);
+    notifications?.toasts.addSuccess('Workspace ID copied');
```
Comment (AMoo-Miki) [inline src/plugins/workspace/public/components/workspace_list/index.tsx:225]: The message is not clear and also should use i18n

File: src/plugins/workspace/public/components/workspace_list/index.tsx [hunk 6]
```diff
@@ -198,6 +234,22 @@ export const WorkspaceList = ({ registeredUseCases$ }: WorkspaceListProps) => {
...
+  const handleSetDefaultWorkspace = useCallback(
+    async (item: WorkspaceAttribute) => {
+      const set = await uiSettings?.set(DEFAULT_WORKSPACE, item.id);
+      if (set) {
+        setDefaultWorkspaceId(item.id);
+        notifications?.toasts.addSuccess(`Default workspace been set to ${item.name}`);
+      } else {
+        // toast
+        notifications?.toasts.addWarning(
+          `Failed to set workspace ${item.name} as default workspace.`
```
Comment (AMoo-Miki) [inline src/plugins/workspace/public/components/workspace_list/index.tsx:246]: i18n plz

File: src/plugins/workspace/public/components/workspace_list/index.tsx [hunk 7]
```diff
@@ -344,6 +405,20 @@ export const WorkspaceList = ({ registeredUseCases$ }: WorkspaceListProps) => {
...
       },
     },
+    {
+      field: 'permissionMode',
+      name: 'Permissions',
+      width: '6%',
+      render: (permissionMode: WorkspaceAttributeWithPermission['permissionMode']) => {
+        return isDashboardAdmin ? (
+          <EuiToolTip position="right" content="You are dashboard admin">
+            <EuiText size="xs">Admin</EuiText>
```
Comment (AMoo-Miki) [inline src/plugins/workspace/public/components/workspace_list/index.tsx:415]: Needs i18n

File: src/plugins/workspace/public/components/workspace_list/index.tsx [hunk 8]
```diff
@@ -361,125 +436,152 @@ export const WorkspaceList = ({ registeredUseCases$ }: WorkspaceListProps) => {
         return renderDataWithMoreBadge(dataSources, 2, item.id, DetailTab.DataSources);
       },
     },
+  ];
+  const allActions = [
+    {
+      name: <EuiText key="copyId">Copy ID</EuiText>,
```
Comment (AMoo-Miki) [inline src/plugins/workspace/public/components/workspace_list/index.tsx:442]: All of these need to use i18n

File: src/plugins/workspace/public/components/workspace_list/index.tsx [hunk 9]
```diff
@@ -361,125 +436,152 @@ export const WorkspaceList = ({ registeredUseCases$ }: WorkspaceListProps) => {
...
+  });
+  const availableColumns = columnsWithoutActions
+    .filter((column) => !excludedColumns.includes(column.field))
+    .filter((column) => {
+      return (
+        !includedColumnsFields ||
+        includedColumnsFields.length === 0 ||
+        includedColumnsFields.includes(column.field)
+      );
+    })
```
Comment (AMoo-Miki) [inline src/plugins/workspace/public/components/workspace_list/index.tsx:507]: We should combine these for performance

File: src/plugins/workspace/public/components/workspace_list/index.tsx [hunk 10]
```diff
@@ -361,125 +436,152 @@ export const WorkspaceList = ({ registeredUseCases$ }: WorkspaceListProps) => {
...
+      const customizedCol = includedColumns.find((col) => col.field === column.field);
+      return {
+        ...column,
+        ...(customizedCol ? { ...customizedCol } : {}),
+      };
+    });
+
+  const actionColumns = [
     {
       name: 'Actions',
```
Comment (AMoo-Miki) [inline src/plugins/workspace/public/components/workspace_list/index.tsx:518]: i18n plz

File: src/plugins/workspace/public/components/workspace_list/index.tsx [hunk 11]
```diff
@@ -361,125 +436,152 @@ export const WorkspaceList = ({ registeredUseCases$ }: WorkspaceListProps) => {
...
+
+  return fullPage ? (
+    <>
+      <EuiPage paddingSize="m">
+        <HeaderControl
+          controls={[
+            {
+              description: i18n.translate('workspace.list.description', {
+                defaultMessage:
+                  'Organize collaborative projects with use-case-specific workspaces.',
```
Comment (AMoo-Miki) [inline src/plugins/workspace/public/components/workspace_list/index.tsx:559]:
```suggestion
                  'Organize collaborative projects with specialized workspaces.',
```

**Comment by AMoo-Miki** [comment]
We should also mark this experimental in the config and changelog to allow for making changes to the API and functionality without causing a breaking change.

**Comment by Hailong-am** [comment]
> We should also mark this experimental in the config and changelog to allow for making changes to the API and functionality without causing a breaking change.

@AMoo-Miki thanks, it should be experimental. I will have the quick follow up PR to address all your comments. have an issue to track https://github.com/opensearch-project/OpenSearch-Dashboards/issues/8072

File: src/plugins/workspace/public/components/workspace_list/index.tsx (hunk 11 above)
Comment (Hailong-am) [reply src/plugins/workspace/public/components/workspace_list/index.tsx:618]: this is comes from UI, so i would like keep it as it is

File: src/plugins/workspace/public/components/workspace_list/index.tsx (hunk 10 above)
Comment (Hailong-am) [reply src/plugins/workspace/public/components/workspace_list/index.tsx]: done

File: src/plugins/workspace/public/components/workspace_list/index.tsx (hunk 9 above)
Comment (Hailong-am) [reply src/plugins/workspace/public/components/workspace_list/index.tsx]: done with combine them into one filter.

File: src/core/server/ui_settings/ui_settings_client.ts [hunk 12]
```diff
@@ -127,25 +165,39 @@ export class UiSettingsClient implements IUiSettingsClient {
     return userProvided;
   }
 
-  async setMany(changes: Record<string, any>) {
-    this.onWriteHook(changes);
-    await this.write({ changes });
+  async setMany(changes: Record<string, any>, scope?: UiSettingScope) {
+    // log changes and scope
+    this.log.debug(`UiSettingsClient.setMany: ${JSON.stringify({ changes, scope })}`);
```
[the remark this answers is not in the document]
Comment (Hailong-am) [reply src/core/server/ui_settings/ui_settings_client.ts]: all dev logging are removed

File: src/core/server/ui_settings/utils.ts (hunk 1 above)
Comment (Hailong-am) [reply src/core/server/ui_settings/utils.ts]: updated to `buildDocIdWithScope`

**Comment by Hailong-am** [comment]
@AMoo-Miki @SuZhou-Joe @ruanyl comments are addressed, please add backport 2.17 label for this PR and close https://github.com/opensearch-project/OpenSearch-Dashboards/pull/8058 this one.

For main, i will have another PR to cherry pick the commit that will not need to backport.

**Review by AMoo-Miki — APPROVED** [review]
Thanks.

**Review by SuZhou-Joe — APPROVED** [review]

**Review by ruanyl — APPROVED** [review]

**Pull request merged**